examples/hybrid/density_current_2dinvariant_rhoe.jl

```diff
@@ -0,0 +1,342 @@
+
+const MSLP = 1e5 # mean sea level pressure
+const grav = 9.8 # gravitational constant
+const R_d = 287.058 # R dry (gas constant / mol mass dry air)
+const γ = 1.4 # heat capacity ratio
+const C_p = R_d * γ / (γ - 1) # heat capacity at constant pressure
+const C_v = R_d / (γ - 1) # heat capacity at constant volume
+const T_0 = 273.16 # triple point temperature
+
+Φ(z) = grav * z
+
+# Reference: https://journals.ametsoc.org/view/journals/mwre/140/4/mwr-d-10-05073.1.xml, Section 5a
+# Prognostic thermodynamic variable: Total Energy 
+function init_dry_density_current_2d(x, z)
+    x_c = 0.0
+    z_c = 3000.0
+    r_c = 1.0
+    x_r = 4000.0
+    z_r = 2000.0
+    θ_b = 300.0
+    θ_c = -15.0
+    cp_d = C_p
+    cv_d = C_v
+    p_0 = MSLP
+    g = grav
+
+    # auxiliary quantities
+    r = sqrt((x - x_c)^2/x_r^2 + (z - z_c)^2/z_r^2)
+    θ_p = r < r_c ? 0.5 * θ_c * (1.0 + cospi(r / r_c)) : 0.0 # potential temperature perturbation
+
+    θ = θ_b + θ_p # potential temperature
+    π_exn = 1.0 - g * z / cp_d / θ # exner function
+    T = π_exn * θ # temperature
+    p = p_0 * π_exn^(cp_d / R_d) # pressure
+    ρ = p / R_d / T # density
+    e = cv_d * (T - T_0) + g * z
+    ρe = ρ * e # total energy
+
+    return (ρ = ρ, ρe = ρe)
+end
+
+# initial conditions
+coords = Fields.coordinate_field(hv_center_space)
+face_coords = Fields.coordinate_field(hv_face_space)
+
+Yc = map(coord -> init_dry_density_current_2d(coord.x, coord.z), coords)
+uₕ = map(_ -> Geometry.Covariant1Vector(0.0), coords)
+w = map(_ -> Geometry.Covariant3Vector(0.0), face_coords)
+Y = Fields.FieldVector(Yc = Yc, uₕ = uₕ, w = w)
+
+energy_0 = sum(Y.Yc.ρe)
+mass_0 = sum(Y.Yc.ρ)
+
+function rhs_invariant!(dY, Y, _, t)
+
+    cρ = Y.Yc.ρ # scalar on centers
+    fw = Y.w # Covariant3Vector on faces
+    cuₕ = Y.uₕ # Covariant1Vector on centers
+    cρe = Y.Yc.ρe
+
+    dρ = dY.Yc.ρ
+    dw = dY.w
+    duₕ = dY.uₕ
+    dρe = dY.Yc.ρe
+    z = coords.z
+
+    # 0) update w at the bottom
+    # fw = -g^31 cuₕ/ g^33
+
+    hdiv = Operators.Divergence()
+    hwdiv = Operators.WeakDivergence()
+    hgrad = Operators.Gradient()
+    hwgrad = Operators.WeakGradient()
+    hcurl = Operators.Curl()
+
+    If2c = Operators.InterpolateF2C()
+    Ic2f = Operators.InterpolateC2F(
+        bottom = Operators.Extrapolate(),
+        top = Operators.Extrapolate(),
+    )
+
+    dρ .= 0 .* cρ
+
+    cw = If2c.(fw)
+    fuₕ = Ic2f.(cuₕ)
+    cuw = Geometry.Covariant13Vector.(cuₕ) .+ Geometry.Covariant13Vector.(cw)
+
+    ce = @. cρe / cρ
+    cI = @. ce - Φ(z) - (norm(cuw)^2) / 2
+    cT = @. cI / C_v + T_0
+    cp = @. cρ * R_d * cT
+
+    h_tot = @. ce + cp / cρ # Total enthalpy at cell centers
+
+    ### HYPERVISCOSITY
+    # 1) compute hyperviscosity coefficients
+    χe = @. dρe = hwdiv(hgrad(h_tot)) # we store χe in dρe
+    χuₕ = @. duₕ = hwgrad(hdiv(cuₕ))
+
+    Spaces.weighted_dss!(dρe)
+    Spaces.weighted_dss!(duₕ)
+
+    κ₄ = 98311.0 # m^4/s
```

Comment:
Do we still need hyperdiffusion if there is second order diffusion?

Reply:
No, we do not necessarily need this - see also the rho-theta example comment .